aiohue/v2/controllers/base.py

```diff
@@ -0,0 +1,234 @@
+        self._logger = bridge.logger.getChild(self.item_type.value)
+        self._subscribers: List[EventSubscriptionType] = []
+
+    @property
+    def items(self) -> List[CLIPResource]:
+        """Return all items for this resource."""
+        return list(self._items.values())
+
+    async def initialize(self, initial_data: List[dict] | None) -> None:
+        """
+        Initialize controller by fetching all items for this resource type from bridge.
+
+        Must be called only once. Any updates will be retrieved by events.
+        Optionally provide the initial data.
+        """
+        if initial_data is None:
+            endpoint = f"clip/v2/resource/{self.item_type.value}"
+            initial_data = await self._bridge.request("get", endpoint)
+        item_count = 0
+        for item in initial_data:
+            if ResourceTypes(item["type"]) != self.item_type:
+                continue
+            item_count += 1
+            res: CLIPResource = parse_clip_resource(item)
+            self._items[res.id] = parse_clip_resource(item)
+        # subscribe to item updates
+        self._bridge.events.subscribe(
+            self._handle_event, resource_filter=self.item_type
+        )
+        self._logger.debug("fetched %s items", item_count)
+
+    def subscribe(
+        self,
+        callback: EventCallBackType,
+        id_filter: str | Tuple[str] | None = None,
+        event_filter: EventType | Tuple[EventType] | None = None,
+    ) -> Callable:
+        """
+        Subscribe to status changes for this resource type.
+
+        Parameters:
+            - `callback` - callback function to call when an event emits.
+            - `id_filter` - Optionally provide a resource ID to filter events for.
+            - `event_filter` - Optionally provide an EventType as filter.
+
+        Returns:
+            function to unsubscribe.
+        """
+        if not isinstance(event_filter, (NoneType, tuple)):
+            event_filter = (event_filter,)
+        if not isinstance(id_filter, (NoneType, tuple)):
+            id_filter = (id_filter,)
+        subscription = (callback, id_filter, event_filter)
+
+        def unsubscribe():
+            self._subscribers.remove(subscription)
+
+        self._subscribers.append(subscription)
+        return unsubscribe
+
+    def get_by_v1_id(self, id: str) -> CLIPResource | None:
+        """Get item by it's legacy V1 id."""
+        return next((item for item in self._items.values() if item.id_v1 == id), None)
+
+    def get_device(self, id: str) -> Device:
+        """Return device the given resource belongs to."""
+        if self.item_type == ResourceTypes.DEVICE:
+            return self[id]
+        for device in self._bridge.devices:
+            for service in device.services:
+                if service.rid == id:
+                    return device
+        # always fallback to bridge device itself
+        return self._bridge.config.bridge_device
```

Comment:
Why not return `None`, and have a different method to get bridge device ID? Otherwise you can't know if a device ID is no longer correct.

Reply:
That makes sense. I did this because resources like groups do not have a device as parent so I could use the same logic but I agree that is messy. I'll adjust this.

Reply:
adjusted this. Side effect is that device-less entities in HA are now no longer attached to the Hue bridge device.
I think that's ok ?